// angular2-text-mask/src/angular2TextMask.js
import { Directive, ElementRef, forwardRef, Input, NgModule, Renderer } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { createTextMaskInputElement } from 'text-mask-core/dist/textMaskCore';
export var MaskedInputDirective = (function () {
    function MaskedInputDirective(renderer, element) {
        this.renderer = renderer;
        this.element = element;
        this.textMaskConfig = {
            mask: '',
            guide: true,
            placeholderChar: '_',
            pipe: undefined,
            keepCharPositions: false,
            onReject: undefined,
            onAccept: undefined
        };
        this._onTouched = function () { };
        this._onChange = function (_) { };
    }
    MaskedInputDirective.prototype.ngOnInit = function () {
        if (this.element.nativeElement.tagName === 'INPUT') {
            // `textMask` directive is used directly on an input element
            this.inputElement = this.element.nativeElement;
        }
        else {
            // `textMask` directive is used on an abstracted input element, `ion-input`, `md-input`, etc
            this.inputElement = this.element.nativeElement.getElementsByTagName('INPUT')[0];
        }
        this.textMaskInputElement = createTextMaskInputElement(Object.assign({ inputElement: this.inputElement }, this.textMaskConfig));
    };
    MaskedInputDirective.prototype.writeValue = function (value) {
        if (this.textMaskInputElement !== undefined) {
            this.textMaskInputElement.update(value);
        }
    };
    MaskedInputDirective.prototype.registerOnChange = function (fn) { this._onChange = fn; };
    MaskedInputDirective.prototype.registerOnTouched = function (fn) { this._onTouched = fn; };
    MaskedInputDirective.prototype.onInput = function ($event) {
        this.textMaskInputElement.update($event.target.value);
        this._onChange($event.target.value);
    };
    MaskedInputDirective.prototype.setDisabledState = function (isDisabled) {
        this.renderer.setElementProperty(this.element.nativeElement, 'disabled', isDisabled);
    };
    MaskedInputDirective.decorators = [
        { type: Directive, args: [{
                    host: {
                        '(input)': 'onInput($event)',
                        '(blur)': '_onTouched()'
                    },
                    selector: '[textMask]',
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return MaskedInputDirective; }),
                            multi: true
                        }]
                },] },
    ];
    /** @nocollapse */
    MaskedInputDirective.ctorParameters = [
        { type: Renderer, },
        { type: ElementRef, },
    ];
    MaskedInputDirective.propDecorators = {
        'textMaskConfig': [{ type: Input, args: ['textMask',] },],
    };
    return MaskedInputDirective;
}());
export var TextMaskModule = (function () {
    function TextMaskModule() {
    }
    TextMaskModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [MaskedInputDirective],
                    exports: [MaskedInputDirective],
                    imports: [CommonModule]
                },] },
    ];
    /** @nocollapse */
    TextMaskModule.ctorParameters = [];
    return TextMaskModule;
}());
export { conformToMask } from 'text-mask-core/dist/textMaskCore';
